### PR TITLE
add pelican reduction to base (no ticket)

### PR DESCRIPTION
**Pull request by geishm-ansto**
**Description of work.**

Added ANSTO Pelican reduction code to the base.

**To test:**

Includes test script in the documentation.
-->
No release notes added as it is only relevant to ansto data files. 
-->

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: docs/source/algorithms/PelicanReduction-v1.rst
Comment (DavidFair) [inline docs/source/algorithms/PelicanReduction-v1.rst:20]: This doc test is good :+1: , could we provide some system tests to validate the algorithm? People typically don't run doc tests to validate the output (and it wouldn't show an error cropping up)

This PR needs system tests before I would personally be comfortable approving it. As we would have no automated way to know if the data produced is still valid, say if something indirectly changed and altered the data in the Mantid framework

A couple of data files (if possible), tests and reference outputs of the various branches (such as separate tests for SofQW and NXSPE) helps prevent subtle changes to the produced data, especially long-term where things might change in the underlying algorithms

**Review by DavidFair — CHANGES_REQUESTED** [review]
There's a lot of comments, but overall the code is pleasant to read. It's well commented, laid out and I can understand what's going on

The bulk of it is readability, such as variable names, magic comments and short variable names. However, we are missing the following:

- A system test of some kind (I'd push you hard to go for more than one though)
- Some release notes so people know that this is available for the next release, as it's substantial and I'm sure there are people waiting for it even if it's ANTSO only (we can make a page or put it into Framework notes)

@ me as soon as you're ready for me to re-review this and I'll be sure to get it tested / approved ASAP so we can try and make the upcoming release

File: Framework/DataHandling/src/LoadPLN.cpp
Comment (geishm-ansto) [reply Framework/DataHandling/src/LoadPLN.cpp:877]: Hi David, thanks for the feedback. I always found string formatting with streams cumbersome and have persisted with snprintf type solutions until C++20 with the <format> library is available. However if it is no longer accepted I'll make the change.

Comment (DavidFair) [reply Framework/DataHandling/src/LoadPLN.cpp:877]: :thinking: I guess ultimately it's getting rid of the manually specified buffer size, I'd be happy to approve with whatever approach if we got rid of that?
Unfortunately, this is the kind of loader where once it's working nobody will look at it for many years so it's likely to be forgotten once C++-20 is available

Comment (geishm-ansto) [reply Framework/DataHandling/src/LoadPLN.cpp]: Changed the code to use string streams.

File: Framework/PythonInterface/plugins/algorithms/PelicanReduction.py
Comment (geishm-ansto) [reply Framework/PythonInterface/plugins/algorithms/PelicanReduction.py]: Removed all future imports.

File: Framework/PythonInterface/plugins/algorithms/PelicanReduction.py
Comment (geishm-ansto) [reply Framework/PythonInterface/plugins/algorithms/PelicanReduction.py]: Changed the variable names to be more  relevant to algorithm.

File: Framework/PythonInterface/plugins/algorithms/PelicanReduction.py
Comment (geishm-ansto) [reply Framework/PythonInterface/plugins/algorithms/PelicanReduction.py:64]: Changed variable names.

File: Framework/PythonInterface/plugins/algorithms/PelicanReduction.py
Comment (geishm-ansto) [reply Framework/PythonInterface/plugins/algorithms/PelicanReduction.py]: Changed variable names.

File: Framework/PythonInterface/plugins/algorithms/PelicanReduction.py
Comment (geishm-ansto) [reply Framework/PythonInterface/plugins/algorithms/PelicanReduction.py]: Added comment that explains function.

File: Framework/PythonInterface/plugins/algorithms/PelicanReduction.py
Comment (geishm-ansto) [reply Framework/PythonInterface/plugins/algorithms/PelicanReduction.py]: Added relevant comment.

File: Framework/PythonInterface/plugins/algorithms/PelicanReduction.py
Comment (geishm-ansto) [reply Framework/PythonInterface/plugins/algorithms/PelicanReduction.py:524]: The raw data files use a non-standard compression algorithm, written sometime ago, that leads to long load times for larger files and a reasonable number of runs. The scratch folder is more like a cache than a short term temp folder which allows users to quickly reprocess the data from a valid intermediate result saved in the folder. The intermediate files need to persist beyond the exceution of the reduction algorithm. A large part of the code in the algorithm deals with the consequences of the slow load times.

File: docs/source/algorithms/PelicanReduction-v1.rst
Comment (geishm-ansto) [reply docs/source/algorithms/PelicanReduction-v1.rst:14]: The description was written by the instrument scientist but he is not back until next week. I'll speak with him on Monday.

File: docs/source/algorithms/PelicanReduction-v1.rst
Comment (geishm-ansto) [reply docs/source/algorithms/PelicanReduction-v1.rst:20]: Added two system tests to handle the SOFQW and NXSPE cases.

On a different issue, the tests failed on two areas completely unrelated to changes here. Could you kick it off again as I don't have the authorization.

Finally, where do I add the release comments about the addition of the Pelican reduction functionality?

**Comment by geishm-ansto** [comment]
Let me talk to the scientist on Monday before merging any changes.

**Comment by geishm-ansto** [comment]
I've spoken with the scientist and he is happy with the workflow addition. Let me know if any further changes are needed.
Thanks.

**Comment by DavidFair** [comment]
Doing a quick grep when the loader was written it was placed into direct_inelastic in `docs/source/release/v4.0.0/direct_inelastic.rst`, what if we create a page at `docs/source/release/v6.1.0/direct_inelastic` (feel free to use the v4.0 page as a template) with it in?

We have someone go through all the release notes and split them up / sort them before the release anyway so if it's not in the optimal place it will get combined with the rest?

Also thanks for doing those changes, I'm doing a re-review but currently this LGTM

File: Testing/SystemTests/tests/framework/PelicanReductionTest.py
Comment (DavidFair) [inline Testing/SystemTests/tests/framework/PelicanReductionTest.py:3]: Can you bump this copyright year to 2021 whilst you add release notes? If it gets forgotten though I won't hold up the PR for something this minor

Comment (geishm-ansto) [reply Testing/SystemTests/tests/framework/PelicanReductionTest.py]: Added release note and updated the license details.

**Review by DavidFair — APPROVED** [review]
LGTM, lets :shipit:

**Pull request merged**
